toe/src/urbit/app/toe.hoon

```diff
@@ -0,0 +1,921 @@
+  --
+::  +view
+::  TODO: please make me into a %view app!
+::
+++  view
+  |%
+  ::  TODO: move other view updates here
+  ::
+  ++  updates
+    |%
+    ++  restart
+      ^-  (list card)
+      :~  (send:co mor+all-effects:co)
+          (send:fe [%status s+'null']~)
+      ==
+    ::
+    ++  receive
+      |=  guest=@p
+      =/  room-name=tape  (cite:title guest)
+      ?^  rooms
+        (notify guest)
+      :~  (send:fe ~[[%message s+(crip room-name)] [%status s+'confirm']])
+          =,  co
+          (send (prompt "{confirm}{room-name}? (Y/N) | "))
+      ==
+    ::
+    ++  confirm
+      =/  next=@p  guest:next:room-core
+      :~  (confirm:updates:fe next)
+          (confirm:updates:co next)
+      ==
+    ::
+    ++  playing
+      |=  move=toe-turno
+      ^-  (list card)
+      =/  [ze=@p who=@p]   =,  current:room-core
+        [ship who]
+      :~  ::  If the move originated in the frontend, this is redundant
+          ::  so this move is ignored on the frontend (via %next)
+          ::
+          %-  send:fe  ^-  (list [@t json])
+          :~  [%status s+'play']
+              [%next s+(scot %p who)]
+              [%stone s+`@t`stone.per.move]
+              [%move a+~[n+(scot %u -.spo.move) n+(scot %u +.spo.move)]]
+          ==
+        ::
+          =,  co
+          %-  send  ^-  sole-effect
+          :~  %mor
+              grid
+              %-  prompt
+              %-  dial
+              [ze stones:room-core ?:(=(who me) "<-" "->")]
+      ==  ==
+    ::
+    ++  wins
+      |=  [out=outcome move=toe-turno winner=@p]
+      ^-  (list card)
+      ~[(wins:updates:fe out move winner) (wins:updates:co [out winner])]
+    ::
+    ++  notify
+      |=  guest=@p
+      :_  ~
+      (send:co klr+~[[[```%b] " [ {<guest>} wants to play ]"]])
+    ::
+    ++  not-allowed
+      :~  (send:co frowned-upon)
+        ::
+          %-  send:fe
+          :~  [%status s+'error']
+              [%message (tape:enjs:format txt:frowned-upon)]
+      ==  ==
+    --
+  ::  %fe: frontend
+  ::
+  ++  fe
+    |%
+    ::  +handle-http-request: serve pages from file system based on URl path
+    ::
+    ++  handle-http-request
+      |=  =inbound-request:eyre
+      ^-  simple-payload:http
+      =/  request-line  (parse-request-line url.request.inbound-request)
+      ~&  request-line
+      ?+  request-line  not-found:gen
+            [[[~ %js] [%'~toe' %js %tile ~]] ~]
+          (js-response:gen tile-js)
+      ==
+    ::
+    ++  poke-json
+      |=  jon=json
+      ^-  (quip card _state)
+      ~&  jon
+      ?.  ?=(%o -.jon)
+        ::  ignores non-object json
+        ::
+        [~ state]
+      =/  object=(map @t json)  +.jon
+      =/  data=json  (~(got by object) 'data')
+      =-  ~(action game-loop (tufa -))
+      ^-  (list @c)
+      ?+    -.data  !!
+        %a  =-  (tuba "{-<}/{->}")
+            =-  :-  (scow %u (snag 0 -))
+                (scow %u (snag 1 -))
+            ((ar:dejs ni:dejs) data)
```

Comment:
You're using `=-` too heavily again. (;

```hoon
%~  action  game-loop
%-  tufa
...
```

Also aren't these `tufa` and `tuba` calls cancelling each other out?

Reply:
This is one of those cases where, first time I looked at hoon code and so this type of construct `=-` I pretty much had no idea what was going on, and somehow now I'm addicted to it :D

Reply:
This is one of those cases where, first time I looked at hoon code and so this type of construct `=-` I pretty much had no idea what was going on, and somehow now I'm addicted to it :D

Reply:
Fix your addiction! (^:

And just to make sure you're aware of it: `=;` exists, in case you want to name things for extra clarity.